### ISSUE TITLE
The test results

### ISSUE THREAD

**Issue by ooobsidian** [philipperemy/deep-speaker#57]
Hello, after training, I used the checkpoint with the best training effect. According to the test sample in readme.md, I changed it slightly and tested 100 speakers outside the training set for cross-validation. Resulting EER has reached ~23% ! Do you have any thoughts on this, is it not suitable for testing data outside the training set?

**Comment by ooobsidian** [comment]
@philipperemy Can you help me please? :(

**Comment by philipperemy** [comment]
What was the EER you had on the training set?

I had `f-measure = 0.849, true positive rate = 0.798, accuracy = 0.997, equal error rate = 0.025` on LibriSpeech.

**Comment by ooobsidian** [comment]
I use `export CUDA_VISIBLE_DEVICES=0; python cli.py test-model --working_dir .deep-speaker-wd/test/ --checkpoint_file checkpoints-triplets/ResCNN_checkpoint_331.h5` , had `2020-05-23 11:29:04,274 - INFO - f-measure = 0.841, true positive rate = 0.816, accuracy = 0.997, equal error rate = 0.024` , but I think it regarded training set as test set, so it has a good performance.

**Comment by philipperemy** [comment]
equal error rate = 0.024 means 2.4% is it's very good!

**Comment by ooobsidian** [comment]
I don't know what your test method is, can you tell me? In my own test file, each speaker has two registered sounds, one test sound. Average the embedding of the two register sounds to get an embedding, and then cross-validation, I get EER is 0.23.

**Comment by ooobsidian** [comment]
My cross validation file content is follow:
```python
import random
import os
import numpy as np
import tensorflow as tf
from tqdm import tqdm
from audio import read_mfcc
from batcher import sample_from_mfcc
from constants import SAMPLE_RATE, NUM_FRAMES
from conv_models import DeepSpeakerModel
from sklearn.metrics.pairwise import cosine_similarity
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
# Reproducible results.
np.random.seed(123)
random.seed(123)

def load_model():
    # Define the model here.
    model = DeepSpeakerModel()
    # Load the checkpoint.
    model.m.load_weights(
        '/data/source/deep-speaker/checkpoints-triplets/ResCNN_checkpoint_331.h5',
        by_name=True)
    return model


def get_per_speaker_embedding(filename, model):
    mfcc = sample_from_mfcc(read_mfcc(filename, SAMPLE_RATE), NUM_FRAMES)
    embedding = model.m.predict(np.expand_dims(mfcc,
                                               axis=0)).flatten()  # (512,)
    return embedding


def embedding_average(path, model):
    audio_list = os.listdir(path)
    sum = [0 for _ in range(512)]
    for idx, audio in enumerate(audio_list):
        tmp = get_per_speaker_embedding(path + '/' + audio, model)
        if idx == 0:
            sum += tmp
        else:
            sum = (sum + tmp) / 2.0
    return sum


def embedding_concatenate(path, model):
    audio_list = os.listdir(path)
    sum = [0 for _ in range(512)]
    for idx, audio in enumerate(audio_list):
        tmp = get_per_speaker_embedding(path + '/' + audio, model)
        if idx == 0:
            sum += tmp
        else:
            sum = np.concatenate((sum, tmp))
    return sum


def get_register(
        model,
        base_path='/data/source/cross_validation_test/data_100/register/',
        out_path='register_emb/'):
    register_files = os.listdir(base_path)
    if not os.path.exists(out_path):
        os.mkdir(out_path)
    for spk in tqdm(register_files, desc="get_register"):
        # ====== 1
        emb = embedding_average(base_path + spk, model)
        # ====== 2
        # emb = embedding_concatenate(base_path + spk,model)
        # print('register shape:', emb.size)
        if not os.path.exists(out_path + spk):
            os.mkdir(out_path + spk)
        target_filename = out_path + spk + '/' + spk + '.npy'
        np.save(target_filename, emb)


def get_validate(
        model,
        base_path='/data/source/cross_validation_test/data_100/verify/',
        out_path='validate_emb/'):
    spk_list = os.listdir(base_path)
    if not os.path.exists(out_path):
        os.mkdir(out_path)
    for spk in tqdm(spk_list, desc="get_validate"):
        tmp_list = os.listdir(base_path + spk)
        for i in tmp_list:
            emb = get_per_speaker_embedding(base_path + spk + '/' + i, model)
            # ====== 1
            # ====== 2
            # emb = np.concatenate((emb, emb))
            # print('validate shape:', emb.shape)
            if not os.path.exists(out_path + spk):
                os.mkdir(out_path + spk)
            target_filename = out_path + spk + '/' + i.split('.')[0] + '.npy'
            np.save(target_filename, emb)


def cross_validation():
    AUDIO_DIR = 'validate_emb/'
    REGISTER_DIR = 'register_emb/'
    spk_list = os.listdir(AUDIO_DIR)
    register_list = os.listdir(REGISTER_DIR)
    print('spk_list: ', spk_list)
    print('register_list: ', register_list)
    ans = ''
    for spk1 in tqdm(spk_list, desc="cross_validation"):
        test_files = os.listdir(AUDIO_DIR + spk1)
        for i in test_files:
            ans = ''
            emb1 = np.load(AUDIO_DIR + spk1 + '/' + i)
            for spk2 in register_list:
                verify_files = os.listdir(REGISTER_DIR + spk2)
                for j in verify_files:
                    emb2 = np.load(REGISTER_DIR + spk2 + '/' + j)
                    ans += str(1 -
                               cosine_similarity([emb1, emb2])[0][1]) + '\t'
            with open('ans.txt', 'a') as f:
                f.write(ans + '\n')


def as_num(x):
    y = '{:.15f}'.format(x)
    return float(y)


if __name__ == '__main__':
    model = load_model()
    get_register(model=model)
    get_validate(model=model)
    cross_validation()
```

**Comment by philipperemy** [comment]
This is the function I use to evaluate a speaker: https://github.com/philipperemy/deep-speaker/blob/ad6deae705f6a05067ce2cbdddb061d358823b72/test.py#L28.

Each speaker is checked another utterance of the same speaker and against 99 negative speakers.

**Comment by philipperemy** [comment]
```python
if idx == 0:
            sum += tmp
        else:
            sum = (sum + tmp) / 2.0
```
Here it seems sum is converging to 0. Be careful :)

**Comment by ooobsidian** [comment]
thanks, I will check it.

**Comment by philipperemy** [comment]
No problem! It's explained in the Deep Speaker paper if the code is not clear!

**Comment by ooobsidian** [comment]
> ```python
> if idx == 0:
>             sum += tmp
>         else:
>             sum = (sum + tmp) / 2.0
> ```
> 
> Here it seems sum is converging to 0. Be careful :)

I don't know what your mean. I averaged the two embedding by this method, and I checked the averaged value and it was correct. Will this method cause errors? Hope to get your answer.

**Comment by ooobsidian** [comment]
@philipperemy Are you free? I'm sorry to bother you.

**Comment by philipperemy** [comment]
@ooobsidian you're doing incremental averaging the wrong way I think: https://math.stackexchange.com/questions/106700/incremental-averageing

**Comment by philipperemy** [comment]
Just append() and np.mean() at the end.

**Comment by philipperemy** [comment]
something like this:
```python
def embedding_average(path, model):
    audio_list = os.listdir(path)
    sum = []
    for idx, audio in enumerate(audio_list):
        tmp = get_per_speaker_embedding(path + '/' + audio, model)
        sum.append(tmp)
    return np.mean(sum)
```

**Comment by ooobsidian** [comment]
Then do you think there are any errors in other parts, such as extracting embeddings. I'll test it using your method.

**Comment by ooobsidian** [comment]
The result of `np.mean(sum)` is a number, not an array. So I think there is something wrong with this method.

> something like this:
> 
> ```python
> def embedding_average(path, model):
>     audio_list = os.listdir(path)
>     sum = []
>     for idx, audio in enumerate(audio_list):
>         tmp = get_per_speaker_embedding(path + '/' + audio, model)
>         sum.append(tmp)
>     return np.mean(sum)
> ```

**Comment by ooobsidian** [comment]
I would like to know how to use the method you provided to test speakers outside the training set. I needed two registered audios and one test audio for each speaker. How should the test program be modified？

**Comment by philipperemy** [comment]
@ooobsidian sorry np.mean(..., axis=0)

**Comment by philipperemy** [comment]
Follow this example: https://github.com/philipperemy/deep-speaker#test-instruction-using-pretrained-model

**Comment by ooobsidian** [comment]
@philipperemy I changed `np.mean(..., axis=0)`, and it's eer is ~23%. And I refer the example Follow this example: https://github.com/philipperemy/deep-speaker#test-instruction-using-pretrained-model.
Here is my frr and far curve:
![](https://pic.alphalrx.cn/20200525221622.png)
I hope you can explain some of this please, thank you.

**Comment by philipperemy** [comment]
@ooobsidian can you run your script (to generate the plot FAR FRR) on the data you trained your model on? Let's make sure that your evaluation script is correct (first of all).

**Comment by ooobsidian** [comment]
@philipperemy I chose 100 speakers in the training set as the test data. There are 30 registered audios for each speaker, and the audios in the training set are very short, basically only 1-2 seconds. In my previous test data, each speaker had two registered audios, each of audio has 20-30 seconds. So the two experiments are basically the same. And I generated the plot FAR FRR, which is follow:
![](https://pic.alphalrx.cn/20200526151032.png)
It's EER = 0.02.

**Comment by philipperemy** [comment]
Hum it's weird. I can't really explain. What I can say is that I trained all the speakers for the training-testing set. 20% of the utterances were put in the test set. But training and testing share the same speakers. Is your test set really different than your training set?

**Comment by ooobsidian** [comment]
Yes, the speakers in the test set I used were not in the training set at all, that is to say, the computer never saw the speaker in the test set. So your system can't distinguish speakers outside of the training set?

**Comment by philipperemy** [comment]
I've never tried but I think it should be able to generalize well because my model was trained on 2500 speakers.

**Comment by ooobsidian** [comment]
So after my attempts, it seems that the system is unable to distinguish the speakers who are not in the training set very well.

**Comment by philipperemy** [comment]
Did you try on a similar dataset such as VCTK Corpus? The datasets of LibriSpeech are clear speech. If your testing set is too far from that (for example noisy env), then it can maybe explain that.

**Comment by ooobsidian** [comment]
So I did it again with some of the speakers in the LibriSpeech or VCTK Corpus. Indeed, there was noise in my test data set.

**Comment by philipperemy** [comment]
Ok so I know from experience that those models work very well in noise free environments. If they haven't been trained with noise, they perform poorly. 

At Baidu, they hired hundreds of people in rooms to generate voice data with different types of sound. Also, the dataset they used contained 250K speakers (which is big!).

**Comment by ooobsidian** [comment]
@philipperemy I tested on VCTK Corpus. I selected 109 speakers, and EER is 3.67%, it's good. So noise in audios is a very important factor.

**Comment by philipperemy** [comment]
@ooobsidian I think we found the reason. EER of 3.67% is good especially this dataset is really out of sample (different speakers, different conditions, but still no noise).
Best is to gather a large noisy dataset and train the model. But I guess, only Baidu can do that...

**Comment by ooobsidian** [comment]
@philipperemy Thank you for your tireless work on this question!

**Comment by philipperemy** [comment]
@ooobsidian no problem!
